File: frontend/lib/api/budget.ts
import { get, put, ApiError } from '@/lib/api/client';

export type BudgetResponse = {
  id: number;
  user_id: number;
  budget_year: number;
  budget_month: number;
  monthly_budget: number;
  created_at: string;
  updated_at: string;
};

export type BudgetSetRequest = {
  budget_year: number;
  budget_month: number;
  monthly_budget: number;
};

/** 指定年月の予算を取得（未設定時は null を返す） */
export async function getBudget(year: number, month: number): Promise<BudgetResponse | null> {
  try {
    return await get<BudgetResponse>(`/api/budget?year=${year}&month=${month}`);
  } catch (error: unknown) {
    if (error instanceof ApiError && error.status === 404) {
      return null;
    }
    throw error;
  }
}

/** 予算を設定（作成 or 更新） */
export async function setBudget(data: BudgetSetRequest): Promise<BudgetResponse> {
  return await put<BudgetResponse>('/api/budget', data);
}

File: backend/app/schemas/budget.py
"""予算設定API用スキーマ"""
from datetime import datetime
from pydantic import BaseModel, Field


class BudgetSetRequest(BaseModel):
    """予算設定リクエスト（作成・更新兼用）"""
    budget_year: int = Field(..., ge=2020, le=2100, description="年")
    budget_month: int = Field(..., ge=1, le=12, description="月")
    monthly_budget: int = Field(..., gt=0, le=99999999, description="月の予算額（円）")


class BudgetResponse(BaseModel):
    """予算レスポンス"""
    id: int
    user_id: int
    budget_year: int
    budget_month: int
    monthly_budget: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
